package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,587 @@
+        RuntimeError
+            when a unit is not recognized by MDAnalysis
+        ValueError
+            when ``n_atoms`` changes values between timesteps
+        ValueError
+            when ``convert_units=True`` but the H5MD file contains no units
+        ValueError
+            when dimension of unitcell is not 3
+        NoDataError
+            when the H5MD file has no 'position', 'velocity', or
+            'force' group
+
+        """
+        if not HAS_H5PY:
+            raise RuntimeError("Please install h5py")
+        super(H5MDReader, self).__init__(filename, **kwargs)
+        self.filename = filename
+        self.convert_units = convert_units
+
+        # if comm is in *kwargs, driver must be 'mpio' and file will be
+        # opened with parallel h5py/hdf5 enabled
+        self._driver = kwargs.get('driver', None)
+        self._comm = kwargs.get('comm', None)
+
+        self.open_trajectory()
+        if self._particle_group['box'].attrs['dimension'] != 3:
+            raise ValueError("MDAnalysis only supports 3-dimensional"
+                             " simulation boxes")
+
+        # _has dictionary used for checking whether h5md file has
+        # 'position', 'velocity', or 'force' groups in the file
+        self._has = {name: name in self._particle_group for
+                     name in ('position', 'velocity', 'force')}
+
+        # Gets n_atoms from first available group
+        for name, value in self._has.items():
+            if value:
+                self.n_atoms = self._particle_group[name]['value'].shape[1]
+                break
+        else:
+            raise NoDataError("Provide at least a position, velocity"
+                              " or force group in the h5md file.")
+
+        self.ts = self._Timestep(self.n_atoms,
+                                 positions=self.has_positions,
+                                 velocities=self.has_velocities,
+                                 forces=self.has_forces,
+                                 **self._ts_kwargs)
+
+        self.units = {'time': None,
+                      'length': None,
+                      'velocity': None,
+                      'force': None}
+        self._set_translated_units()  # fills units dictionary
+        self._read_next_timestep()
+
+    @staticmethod
+    def _format_hint(thing):
+        """Can this Reader read *thing*"""
+        # nb, filename strings can still get passed through if
+        # format='H5MD' is used
+        return HAS_H5PY and isinstance(thing, h5py.File)
+
+    def open_trajectory(self):
+        """opens the trajectory file using h5py library"""
+        self._frame = -1
+        if isinstance(self.filename, h5py.File):
+            self._file = self.filename
+            self._driver = self._file.driver
+        else:
+            if self._comm is not None:
+                # can only pass comm argument to h5py.File if driver='mpio'
+                assert self._driver == 'mpio'
```

Comment:
If you `assert` here then you must have made sure _before_ that the assertion is true. Don't use `assert` for input checking: The message confuses users and `assert` can be optimized away. You have three choices
1. leave the `assert` here and add an explicit test in `__init__` when you process `comm` and `driver`
2. replace the `assert` here with raising an exception
3. remove the `assert` and let `h5py.File` raise an exeption

If (3) is not giving a clear error message then you should raise one yourself. i.e., (1) or (2). My preference is (1) – fail early.

Reply:
I went with (1) as the error given by h5py is a bit confusing